README.md
# Larry 🐦
[![Go](https://github.com/ezeoleaf/larry/actions/workflows/go.yml/badge.svg?branch=main)](https://github.com/ezeoleaf/larry/actions/workflows/go.yml)
[![Coverage Status](https://coveralls.io/repos/github/ezeoleaf/larry/badge.svg?branch=main)](https://coveralls.io/github/ezeoleaf/larry?branch=main)
[![Go Report Card](https://goreportcard.com/badge/github.com/ezeoleaf/larry)](https://goreportcard.com/report/github.com/ezeoleaf/larry)
[![MIT License](https://img.shields.io/github/license/ezeoleaf/larry?style=flat-square)](https://github.com/ezeoleaf/larry/blob/main/LICENSE)
[![Contribute with Gitpod](https://img.shields.io/badge/Contribute%20with-Gitpod-908a85?logo=gitpod)](https://gitpod.io/#https://github.com/ezeoleaf/larry)

Larry is a Golang cli bot that tweets random Github repositories.

## Disclaimer

I hold no liability for what you do with this bot or what happens to you by using this bot. Abusing this bot *can* get you banned from Twitter, so make sure to read up on [proper usage](https://support.twitter.com/articles/76915-automation-rules-and-best-practices) of the Twitter API.

## Running bots

- [GolangRepos](https://twitter.com/GolangRepos): Tweets repositories from Github that contain the "golang" topic
- [RustRepos](https://twitter.com/RustRepos): Tweets repositories from Github that contain the "rust" topic
- [MLRepositories](https://twitter.com/MLRepositories): Tweets repositories from Github that contain the "machine-learning" topic
- [CryptoRepos](https://twitter.com/CryptoRepos): Tweets repositories from Github that contain the "crypto" topic

## Installation

You can install Larry by cloning the repo and using `go install`

```bash
git clone https://github.com/ezeoleaf/larry.git
cd larry/cmd/larry
go install
```

You can also use make for building the project and generating an executable:
```bash
git clone https://github.com/ezeoleaf/larry.git
cd larry
make build
```

Or you can just run it on the go
```bash
git clone https://github.com/ezeoleaf/larry.git
cd larry
go run . [options]
```

## Usage

### Configuring the bot

Before running the bot, you must first set it up so it can connect to Github and the publisher's API.

To do this, you will need to setup the following environment variables:

- GITHUB_ACCESS_TOKEN

For generating Github access token you can follow this [guide](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)

### Providers and Publishers

For information on publishers and providers click [here](PublishersAndProviders.md)

### Running the bot

To run the bot, you have two ways.

If you have installed the bot, you can run it using
  `larry [options]`

If you want to run it without installing it globally you can use
  `go run . [options]`

Example:

`larry -h`

As a response you will see the entire options available

```
NAME:
   Larry - Bot that publishes information from providers to different publishers

USAGE:
   larry [global options] command [command options] [arguments...]

AUTHORS:
   @ezeoleaf <ezeoleaf@gmail.com>
   @beesaferoot <hikenike6@gmail.com>
   @shubhcoder
   @kannav02
   @siddhant-k-code <siddhantkhare2694@gmail.com>
   @savagedev

COMMANDS:
   help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --topic value, -t value          topic for searching repos
   --lang value, -l value           language for searching repos
   --time value, -x value           periodicity of tweet in minutes (default: 15)
   --cache value, -r value          size of cache for no repeating repositories (default: 50)
   --hashtag value, --ht value      list of comma separated hashtags
   --tweet-language, --tl           bool for allowing twetting the language of the repo (default: false)
   --safe-mode, --sf                bool for safe mode. If safe mode is enabled, no repository is published (default: false)
   --provider value, --pr value     provider where publishable content comes from (default: "github")
   --publisher value, --pub value   list of comma separated publishers (default: "twitter")
   --content-file value, --cf value file containing content to publish
   --skip-csv-header, --sh          bool to skip CSV file header. If true, then first record of CSV file is skipped (default: false)
   --blacklist value, --bl value    optional file containing blacklisted repository Ids
   --help, -h                       show help (default: false)
```

For running the bot, the command will depend on whatever you want to tweet, but, for tweeting about React repositories every 30 minutes, you could use

&nbsp;&nbsp;`larry --topic react --time 30 --safe-mode`

For running the bot for Rust tweets every 15 minutes

&nbsp;&nbsp;`larry --lang rust --time 15`

For running the bot for Golang every 15 minutes and specifying a blacklist file named blacklist.txt

&nbsp;&nbsp;`larry --topic golang --time 15 --blacklist ./blacklist.txt`

For running the bot every 60 minutes using the "contentfile" provider and JSON file for content

&nbsp;&nbsp;`larry --time 60 --provider contentfile --content-file ./content.json`

For running the bot every 60 minutes using the "contentfile" provider to read CSV file for content and skipping the header record

&nbsp;&nbsp;`larry --time 60 --provider contentfile --content-file ./content.csv --skip-csv-header`


## Content Files

The `contentfile` provider serves content from CSV and JSON files.

### JSON Content File

When the `contentfile` provider receives a `content-file` filename with a `.json` extension, the provider serves random content from the JSON file. This file consists of an array of objects in the following format. ExtraData is an array of strings.

```json
[{
   "Title": "larry",
   "Subtitle": "Larry 🐦 is a bot generator that publishes random content from different providers built in Go",
   "URL": "github.com/ezeoleaf/larry",
   "ExtraData": ["68", "ezeoleaf", "golang"]
}]
```

### CSV Content File

When the `contentfile` provider receives a `content-file` filename with a `.csv` extension, the provider serves random content from the CSV file. Each field may or may not be enclosed in double quotes. The ExtraData strings start at field 4 of the record and a record can contain any number of elements. 

The following file has one record with three ExtraData strings.

```csv
The title,The subtitle,URL,ExtraString1,"ExtraString2,has comma",ExtraString3
```

An example CSV file with a header record followed by one record:

```csv
Title,Subtitle,URL,Stars,Author,Language,ExtraData1,ExtraData2,ExtraData3
larry,Larry 🐦 is a bot generator that publishes random content from different providers built in Go,github.com/ezeoleaf/larry,68,ezeoleaf,golang
```

Note: Every record in the CSV file, including the header record, must have the same number of fields otherwise an error will occur. This means if the records will have a variable number of ExtraData fields, each record having fewer than the maximum ExtraData fields must include empty ExtraData fields to match the maximum.

## Blacklist File

### Github Provider

For the `github` provider, the optional blacklist file consists of numeric GitHub repository IDs to exclude from the publishing process. These IDs can be found on the GitHub repository page source in the meta tag `octolytics-dimension-repository_id`.

An example blacklist file containing GitHub repository IDs. The file can contain comments using the # character, everything on the line after this character is ignored.

```csv
# Blacklisted repositories
123 # description of the respository
456
```

### Contentfile Provider

For the `contentfile` provider, the optional blacklist file consists of content titles to exclude from the publishing process.

## Have questions? Need help with the bot?

If you're having issues with or have questions about the bot, [file an issue](https://github.com/ezeoleaf/larry/issues) in this repository so anyone can get back to you.

Or feel free to contact me <ezeoleaf@gmail.com> :)

## Star History

[![Star History Chart](https://api.star-history.com/svg?repos=ezeoleaf/larry&type=Date)](https://star-history.com/#ezeoleaf/larry)

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/H2H47X7QW)


CHANGELOG.md
# Changelog

## [0.10.0](https://github.com/ezeoleaf/larry/compare/0.9.0...v0.10.0) (2023-05-13)


### Features

* add release please ([0849781](https://github.com/ezeoleaf/larry/commit/0849781b7a92f687c305face6d95537961076df5))


### Bug Fixes

* coveralls ([11c2020](https://github.com/ezeoleaf/larry/commit/11c20200f4bb506b19e0b743a3682d348209c7b7))
* minor space in tweet ([92a588e](https://github.com/ezeoleaf/larry/commit/92a588e9847c14cde50f186f2ced6dd7781ccb12))
* release please action ([e4aad30](https://github.com/ezeoleaf/larry/commit/e4aad30f0e035f6e60e673d38d046f3e88e3e1d9))


### Miscellaneous

* clean-up docs ([1a9dcff](https://github.com/ezeoleaf/larry/commit/1a9dcff3b68a44fbb9b817bce5786ad2024f11e8))
* switch to goreleaser for building ([696021a](https://github.com/ezeoleaf/larry/commit/696021a4774380f1ce43d3abfeb02ff1bbed023c))
* update workflow ([627f101](https://github.com/ezeoleaf/larry/commit/627f101827530df992cf44fd3b5486d42010f339))


larry/flags.go
package larry

import (
	cfg "github.com/ezeoleaf/larry/config"
	"github.com/urfave/cli/v2"
)

// GetFlags returns a list of flags used the application
func GetFlags(cfg *cfg.Config) []cli.Flag {
	return []cli.Flag{
		&cli.StringFlag{
			Name:        "topic",
			Aliases:     []string{"t"},
			Value:       "",
			Usage:       "topic for searching repos",
			Destination: &cfg.Topic,
		},
		&cli.StringFlag{
			Name:        "lang",
			Aliases:     []string{"l"},
			Value:       "",
			Usage:       "language for searching repos",
			Destination: &cfg.Language,
		},
		&cli.IntFlag{
			Name:        "time",
			Aliases:     []string{"x"},
			Value:       15,
			Usage:       "periodicity of tweet in minutes",
			Destination: &cfg.Periodicity,
		},
		&cli.IntFlag{
			Name:        "cache",
			Aliases:     []string{"r"},
			Value:       50,
			Usage:       "size of cache for no repeating repositories",
			Destination: &cfg.CacheSize,
		},
		&cli.StringFlag{
			Name:        "hashtag",
			Aliases:     []string{"ht"},
			Value:       "",
			Usage:       "list of comma separated hashtags",
			Destination: &cfg.Hashtags,
		},
		&cli.BoolFlag{
			Name:        "tweet-language",
			Aliases:     []string{"tl"},
			Value:       false,
			Usage:       "bool for allowing twetting the language of the repo",
			Destination: &cfg.TweetLanguage,
		},
		&cli.BoolFlag{
			Name:        "safe-mode",
			Aliases:     []string{"sf"},
			Value:       false,
			Usage:       "bool for safe mode. If safe mode is enabled, no repository is published",
			Destination: &cfg.SafeMode,
		},
		&cli.StringFlag{
			Name:        "provider",
			Aliases:     []string{"pr"},
			Value:       "github",
			Usage:       "provider where publishable content comes from",
			Destination: &cfg.Provider,
		},
		&cli.StringFlag{
			Name:        "publisher",
			Aliases:     []string{"pub"},
			Value:       "twitter",
			Usage:       "list of comma separared publishers",
			Destination: &cfg.Publishers,
		},
		&cli.StringFlag{
			Name:        "blacklist",
			Aliases:     []string{"bl"},
			Value:       "",
			Usage:       "optional file containing blacklisted repository Ids",
			Destination: &cfg.BlacklistFile,
		},
		&cli.StringFlag{
			Name:        "content-file",
			Aliases:     []string{"cf"},
			Value:       "",
			Usage:       "file containing content to publish",
			Destination: &cfg.ContentFile,
		},
		&cli.BoolFlag{
			Name:        "skip-csv-header",
			Aliases:     []string{"sh"},
			Value:       false,
			Usage:       "bool to skip CSV file header. If true, then first record of CSV file is skipped",
			Destination: &cfg.SkipCsvHeader,
		},
	}
}
